### PR TITLE
Fix logical error in arrayUnion and arraySymmetricDifference with a Nullable and a Dynamic argument

### LINKED ISSUE #118961: arrayUnion(Nullable, Dynamic) LOGICAL_ERROR: element type stripped, column not
### Describe what's wrong

**`SELECT arrayUnion([toNullable(1)], ['a'::Dynamic])` and the same shape for `arraySymmetricDifference` now throw `Code: 49 ... Columns are assumed to be of identical types, but they are different in Nullable. (LOGICAL_ERROR)`. The not `Nullable` spelling `arrayUnion([1], ['a'::Dynamic])` returns `[1,'a']` and is asserted by the existing test 03726. Reachable from plain table columns: `SELECT arrayUnion(n, d) FROM t` with `n Array(Nullable(Int64)), d Array(Dynamic)`.**

- **Root cause:** The element types are stripped of `Nullable` unconditionally at :405-408, but the corresponding *column* is unwrapped only inside `if (typeid_cast<const ColumnNullable *>(arg.nested_column))` at :391-398 - i.e. only when the CAST element column is a `ColumnNullable`. With an `Array(Dynamic)` cast target the cast column is a `ColumnDynamic`, so the branch is skipped and `initial_column` stays a `ColumnNullable`, while :418 now labels it with the stripped type `UInt8`. `notEquals` therefore receives a column that contradicts its declared type. Pre-PR the mismatch was unreachable: `isInteger(Nullable(UInt8))` is `WhichDataType(Nullable).isInteger()` = `isInt() || isUInt()` = false ([`src/DataTypes/IDataType.h:409`](https://github.com/ClickHouse/ClickHouse/blob/602ce8d11aa216c/src/DataTypes/IDataType.h#L409), free function at [`src/DataTypes/IDataType.cpp:452`](https://github.com/ClickHouse/ClickHouse/blob/602ce8d11aa216c/src/DataTypes/IDataType.cpp#L452)), so the guard never fired for an `Array(Nullable(T))` argument at all.

<details>
<summary>Analysis details (evidence, affected locations, impact)</summary>

**Why we believe this is a bug:** `FunctionArrayIntersect::executeImpl` (arrayIntersect.cpp:478) -> `prepareArrays` -> the overflow-mask guard at :401-419 -> `callFunctionNotEquals` (:416) -> `ExecutableFunctionDynamicAdaptor::executeImpl` -> `FunctionComparison<NotEqualsOp>::executeGenericIdenticalTypes` -> throw at FunctionsComparison.h:1428. Stack trace from the failing run names frames 4, 9 and 13 of exactly that chain.

**Affected locations:**
- [`src/Functions/array/arrayIntersect.cpp:405`](https://github.com/ClickHouse/ClickHouse/blob/602ce8d11aa216c/src/Functions/array/arrayIntersect.cpp#L405) — PR strips `Nullable` from both element types unconditionally
- [`src/Functions/array/arrayIntersect.cpp:396`](https://github.com/ClickHouse/ClickHouse/blob/602ce8d11aa216c/src/Functions/array/arrayIntersect.cpp#L396) — `initial_column` is unwrapped only when the cast element column is a `ColumnNullable`
- [`src/Functions/array/arrayIntersect.cpp:418`](https://github.com/ClickHouse/ClickHouse/blob/602ce8d11aa216c/src/Functions/array/arrayIntersect.cpp#L418) — still-wrapped `initial_column` handed to `notEquals` with the stripped type
- [`src/Functions/FunctionsComparison.h:1428`](https://github.com/ClickHouse/ClickHouse/blob/602ce8d11aa216c/src/Functions/FunctionsComparison.h#L1428) — paranoid nullability check that turns the mismatch into LOGICAL_ERROR

**Impact:** Any `arrayUnion` / `arraySymmetricDifference` mixing an `Array(Nullable(<integer|Date|DateTime|DateTime64>))` argument with an `Array(Dynamic)` argument fails with LOGICAL_ERROR instead of returning a result - const or non-const, 2 or more arguments, and with real table columns. The server-side `ForcedCriticalErrorsLogger` fires, so this also reddens stateless/fuzzer CI jobs once any test reaches the shape.

</details>

### Does it reproduce on most recent release?

Yes — confirmed on current `master` (commit `602ce8d11aa216c`).

### How to reproduce

[▶ Run on ClickHouse Fiddle](https://fiddle.clickhouse.com/83494462-8540-4a52-8669-9e724fe1933b)

<details>
<summary>Reproducer</summary>

```sql
SELECT arrayUnion([1],  ['a'::Dynamic]);
SELECT arrayUnion([toNullable(1)], ['a'::Dynamic]);
SELECT arraySymmetricDifference([toNullable(1)], ['a'::Dynamic]);
```

</details>

### Expected behavior

```
[1,'a']
[1,'a']
[1,'a']
```

### Error message and/or stacktrace

```
[1,'a']
```

<details>
<summary>Suggested fix</summary>

Unwrap `initial_column` from `ColumnNullable` independently of the cast column, so the column always matches the stripped type - move the inner unwrap out of the `if` at :391, e.g. `if (const auto * initial_nullable = typeid_cast<const ColumnNullable *>(initial_column)) initial_column = &initial_nullable->getNestedColumn();` after :398. Placing it before the `arg.nested_column != initial_column` check at :401 is safe: when no cast happened both sides unwrap to the same object and stay equal.

</details>
<details>
<summary>Additional context</summary>

**Open risks:**
- `arrayIntersect` is spared only because `getMostSubtype` rejects `Dynamic` (NO_COMMON_TYPE). Any cast target whose column is not a `ColumnNullable` reaches the same line - `Variant` and `JSON` are currently unreachable because `getLeastSupertype` refuses them next to a number, so the blast radius today is exactly `Dynamic`.

Found during automated review of [PR #118485](https://github.com/ClickHouse/ClickHouse/pull/118485). Severity P1 · Finding `h_pr118485_001`

</details>

cc @groeneai @PedroTadim (from #118485)

### PR DESCRIPTION
**Pull request by groeneai**
Closes: https://github.com/ClickHouse/ClickHouse/issues/118961
Related: https://github.com/ClickHouse/ClickHouse/pull/118485

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Not required: the defect exists only on master, introduced by the unreleased #118485, so no released version is affected.

### Description

`SELECT arrayUnion([1, NULL], ['a'::Dynamic])` fails with `Columns are assumed to be of identical types, but they are different in Nullable`, and aborts debug and sanitizer builds. `arraySymmetricDifference` behaves the same, and table columns fail identically, not only constants.

`prepareArrays` builds an overflow mask by re-comparing the cast element column against the original with `notEquals`, and it unwraps `ColumnNullable` from the initial column only inside the `if` that tests the cast one. With a `Dynamic` common element type the cast column is a `ColumnDynamic`: `Dynamic` cannot sit inside `Nullable`, and `getLeastSupertype`'s `Dynamic` rule precedes its `Nullable` rule. So that inner unwrap never runs and the initial column keeps its wrapper, while #118485 strips `Nullable` from both declared types, leaving `callFunctionNotEquals` with `UInt8` for a still-`ColumnNullable` column. The `Dynamic` adaptor substitutes its own operand's per-variant column, so the declared types match and `notEquals` enters `executeGenericIdenticalTypes`, whose paranoid `isNullable()` check this argument list violates.

The fix moves that unwrap out of the cast-column `if`, so both operands are unwrapped whenever either is. `initial_column` is read nowhere after the guard block, so the move is observable only where it fails today: with no cast both sides unwrap to the same pointer and the guard is skipped as before. Before #118485 the element type stayed `Nullable(T)` and `isInteger(Nullable(T))` is false, so the branch was unreachable in every release: master only.

Validated both directions: four query shapes and the test fail before and pass after, returning the pre-#118485 output `[NULL,1,'a']`; 50 randomized runs and five neighbour tests pass. `arrayIntersect` and the `Variant` spelling stop at an earlier `NO_COMMON_TYPE`, so they are not carriers. Residual left alone: the mask is all-zero whenever the cast target is `Dynamic`, which predates this defect.

CI provenance: https://s3.amazonaws.com/clickhouse-test-reports/praktika.html?PR=117973&sha=fea6f43a91efed27ff56e9425b728dfda0498c74&name_0=PR&name_1=Stress%20test%20%28arm_tsan%29

---
Workflow [[PR](https://s3.amazonaws.com/clickhouse-test-reports/praktika.html?PR=118966&sha=latest&name_0=PR)]
Sync PR [[sync-upstream/pr/118966](https://github.com/search?q=head%3Async-upstream%2Fpr%2F118966+org%3AClickHouse+type%3Apr&type=pullrequests)]

### Version info
- Merged into: `26.9.1.1272` (included in `26.9` and later)

### CODE REVIEW

**Comment by groeneai** [comment]
<details>
<summary>Internal second-model review: no findings in the code; two notes disclosed</summary>

An independent model re-reviewed this diff cold and returned no findings. My own cold pass raised five,
all in the PR description rather than the source, and I fixed those before publishing.

💡 The moved guard tests `initial_column->isNullable()` and then reference-casts to `ColumnNullable`,
where the sibling unwrap just above uses the pointer form. `isNullable()` is also overridden by
`ColumnSparse`, `ColumnReplicated` and `ColumnConst`, but none of them can be a `ColumnArray`'s data at
that line: `ColumnConst` is unwrapped earlier, a replicated argument cannot pass the two `ColumnArray`
casts that precede it, and `Array` is not sparse-serializable. I kept the two-line move instead of
rewriting the predicate, since the hazard has no reachable input. Happy to switch to the pointer form
if you prefer it.

💡 Whenever the cast target is `Dynamic` the overflow mask is all-zero by construction, so computing it
is wasted work. That predates this defect and I left it alone.

</details>

**Comment by groeneai** [comment]
cc @Avogar @Ergus, could you review this? `prepareArrays` unwrapped `ColumnNullable` from the initial element column only inside the `if` that tests the cast column, so with a `Dynamic` common element type (which cannot sit inside `Nullable`) the overflow-mask `notEquals` received a still-`ColumnNullable` column declared as bare `UInt8`, and its identical-types check raises a logical error. The fix moves that unwrap out of the `if`; master only, since the branch was unreachable for `Nullable` element types before #118485.

File: tests/queries/0_stateless/05104_array_intersect_nullable_overflow.sql [hunk 1]
```diff
@@ -7,3 +7,8 @@ SELECT arrayIntersect([toNullable(257)], [toNullable(1)]);
 SELECT arraySort(arrayIntersect([toNullable(1), toNullable(2)], [toNullable(257), toNullable(2)]));
 SELECT arraySort(arrayIntersect([toLowCardinality(toNullable(1))], [toLowCardinality(toNullable(257))]));
 SELECT arraySort(arrayUnion([1], [toNullable(257)]));
+
+-- A common element type that cannot be `Nullable` leaves the cast element column unwrapped while the
+-- original one is not: the guard must not hand the comparison a column and a type that disagree.
+SELECT arrayUnion([1, NULL], ['a'::Dynamic]);
```
Comment (PedroTadim) [inline tests/queries/0_stateless/05104_array_intersect_nullable_overflow.sql:13]: @groeneai can you add more test cases, such as with other array functions, data types and empty arrays?

Comment (groeneai) [reply tests/queries/0_stateless/05104_array_intersect_nullable_overflow.sql:13]: Added in a22f0d5def86e9008cfd137b1f65fcb8de76f189, each case measured on both a pre-fix master build and this branch:

- `Date`, `DateTime` and `DateTime64` element types, one per guard predicate (the mask branch tests `isInteger`, `isDate`, `isDateTime` and `isDateTime64` separately), spread over `arrayUnion` and `arraySymmetricDifference`
- `LowCardinality(Nullable(UInt8))` elements
- non-constant arguments, so the abort is reached through `executeImpl` and not only through the analyzer's constant folding
- a value present in both arguments, which also pins that the overflow mask stays aligned with the element column it was built from, since the result has to be deduplicated
- an empty array, paired with a non-empty row in the same column

All of them abort on master with `Logical error: 'Columns are assumed to be of identical types, but they are different in Nullable'` and return the expected value here.

On the empty argument specifically: on its own it passes on master too, because the cast side is then a zero-row `ColumnDynamic` with no variants, so the comparison that carries the check is never dispatched. That is why I pair it with a non-empty row in the same column, otherwise the case cannot fail either way.

Shapes I measured and left out, with the reason:

- `arrayIntersect` cannot reach this at all. It takes its cast target from `getMostSubtype`, which rejects a number next to `Dynamic` with `NO_COMMON_TYPE` before `prepareArrays` runs, so `arrayUnion` and `arraySymmetricDifference` are the only two entry points.
- `String`, `Float64` and `Enum8` elements are not in the guard's type list, and `Variant` or `JSON` in place of `Dynamic` has no supertype with a number (`NO_COMMON_TYPE`). All five behave identically with and without the fix, so a case for them cannot fail.
- `Int256` elements, reversed argument order, and a `Date` variant inside the `Dynamic` argument do abort on master and are fixed, but each goes through a guard predicate that an added case already covers. Happy to add them if you want the shapes pinned anyway.

The results are wrapped in `arraySort` because the element order of an `Array(Dynamic)` is variant dependent and order is not what these cases check. The file passes 50 out of 50 randomized runs locally.

**Comment by groeneai** [comment]
### CI finish ledger - a22f0d5def86e9008cfd137b1f65fcb8de76f189

Every failure below has an owner: a fixing PR (mine or external), or a full-effort fix task whose
fixing-PR link will be posted here when it opens.

| Check / test | Reason | Owner / fixing PR |
|---|---|---|
| Integration tests (amd_asan_ubsan, db disk, old analyzer, 1/6) / `Timeout` | job-level session timeout (`session-timeout occurred during test execution`), zero failing tests; 197 rows over 163 pull requests and 14 true-master runs in 30 days | #118384 (mine, open) and #118933 (external, open) |
| Stress test (amd_msan) / `Input nodes size mismatch in dag` (STID 4752-5ad8) | `LOGICAL_ERROR` abort at `JoinExpressionActions:138` reached through `chooseJoinOrder`, on a `Nullable(Nothing)` join key; 11 rows over 11 unrelated pull requests, 0 on master, in 30 days, so not caused by this PR (nothing here touches the join planner) | #119106 (mine, open) |
| Stress test (arm_debug) / `Inconsistent KeyCondition behavior` (STID 5182-6d6b) | debug-only key-analysis canary abort at `MergeTreeDataSelectExecutor.cpp:2550` (`markRangesFromPKRange`); 4 rows over 4 pull requests, 0 on master, in 30 days, and no index or key-analysis code is touched here | a fix task is created (investigating at full effort, fixing-PR link to follow here) |

`Mergeable Check` and `PR` are the praktika aggregators reporting the three rows above, not separate
failures. `CH Inc sync` is green at this head. 148 of 174 check-runs passed, 21 were skipped, and the
workflow finished at 2026-09-10 02:38:06Z.

Session id: cron:our-pr-ci-monitor:20260910-030114

**Review by PedroTadim — APPROVED** [review]

**Comment by PedroTadim** [comment]
@alexey-milovidov the PR is on the queue, why you merged master into it?

**Comment by alexey-milovidov** [comment]
@PedroTadim, it had a failed check.

**Comment by groeneai** [comment]
Fixing PR for the `Inconsistent KeyCondition behavior` row on my CI finish ledger here (STID 5182-6d6b, `Stress test (arm_debug)`, 2026-09-09): #121089.
